### PR TITLE
[d3-array] Update to 2.0 - add quickselect, group, and rollup

### PR DESCRIPTION
**Pull request by Ledragon**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [2.0.0](https://github.com/d3/d3-array/releases/tag/v2.0.0)
- [x] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

Should resolve issue #32162

### CODE REVIEW

**Review by gustavderdrache — CHANGES_REQUESTED** [review]
I'm not convinced the types of `group` and `rollup` are correct.

**Review by denisname — CHANGES_REQUESTED** [review]

**Comment by denisname** [comment]
d3-array 2.0 "accept iterables (not just arrays) where relevant". You have not done that part. This v2 is a big update...

File: types/d3-array/index.d.ts [hunk 1]
```diff
@@ -189,6 +228,22 @@ export function descending(a: Primitive | undefined, b: Primitive | undefined):
 // Transforming Arrays
 // --------------------------------------------------------------------------------------
 
+/**
+ * Groups the specified array of values into a Map from key to array of value.
+ * @param a The array to group.
+ * @param key The key function.
+ */
+export function group<TObject, TKey>(a: ArrayLike<TObject>, key: (value: TObject) => TKey): Map<TKey, TObject>;
```
[the remark this answers is not in the document]
Comment (Ledragon) [reply types/d3-array/index.d.ts]: Not to me: done

File: types/d3-array/index.d.ts [hunk 2]
```diff
@@ -189,6 +228,22 @@ export function descending(a: Primitive | undefined, b: Primitive | undefined):
...
+
+/**
+ * Groups and reduces the specified array of values into a Map from key to value.
+ *
+ * @param a The array to group.
+ * @param reduce The reduce function.
+ * @param key The key function.
+ */
+export function rollup<TObject, TKey>(a: ArrayLike<TObject>, reduce: (value: ArrayLike<TObject>) => number, key: (value: TObject) => TKey): Map<TKey, TObject>;
```
[the remark this answers is not in the document]
Comment (Ledragon) [reply types/d3-array/index.d.ts]: Not to me: done

**Comment by Ledragon** [comment]
Temporary update to see how it goes through CI, for sure not complete, the whole `iterable` stuff is not complete

**Comment by Ledragon** [comment]
Can't dedicate any more time to this ATM, I keep going on this afternoon

**Comment by Ledragon** [comment]
To sum up what has been going on:
* I created the v1 sub folder for d3-array
* I referenced it where relevent
* I had to update a generic return type to get the tests to pass. I don't know why it broke, but it seems to be due to TS 3.3. I opened a review
* I have a branch where all `ArrayLike` have been converted to `Iterable`. I however have a doubt about a couple of methods, whether they have to be changed or not. I opened [an issue](https://github.com/d3/d3-array/issues/91) on d3-array to ask about these

File: types/d3-array/d3-array-tests.ts [hunk 3]
```diff
@@ -608,7 +631,7 @@ const testObject = {
 };
 
 const p1: Array<number | string | Date | number[]> = d3Array.permute(testObject, ['name', 'val', 'when', 'more']);
-const p2: Array<Date | number[]> = d3Array.permute(testObject, ['when', 'more']);
+const p2: Array<number | string | Date | number[]> = d3Array.permute(testObject, ['when', 'more']);
```
Comment (denisname) [inline types/d3-array/d3-array-tests.ts]: Why you changed this?

**Review by denisname — CHANGES_REQUESTED** [review]

File: types/d3-array/d3-array-tests.ts (hunk 3 above)
Comment (denisname) [reply types/d3-array/d3-array-tests.ts]: Also remove all `@deprecated` in v2.

**Comment by Ledragon** [comment]
@denisname I'm not forgetting about this, I hope I get more time to finish it properly soon. 

As for your question, I changed the return type of the test because it fails with Tyepscript 3.3, and I have no clue why

**Pull request closed without merging**